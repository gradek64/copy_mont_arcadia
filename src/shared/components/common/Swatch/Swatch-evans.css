/** @define Swatch */

.Swatch {
  &.is-selected {
    &:before {
      content: '';
      border: $plp-product-swatch-button-after-border;
      z-index:1;
      width: 100%;
      height: 100%;
      display: block;
      position: absolute;
    }
  }
}
