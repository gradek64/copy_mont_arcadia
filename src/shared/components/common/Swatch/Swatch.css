/** @define Swatch */

$SwatchSizeMobile: 22px;
$SwatchSizeDesktop: 28px;
$SwatchMarginMobile: 0 6px;
$SwatchMarginMobile-firstChild: 0 6px 0 0;
$SwatchMarginDesktop: 0 10px;
$SwatchMarginDesktop-firstChild: 0 10px 0 0;

.Swatch {
  margin: $SwatchMarginMobile;
  position: relative;
  box-sizing: border-box;
  width: $SwatchSizeMobile;
  height: $SwatchSizeMobile;
  border-radius: $plp-product-swatch-radius;
  transform: translate3d(0, 0, 0);

  &:first-child {
    margin: $SwatchMarginMobile-firstChild;
  }

  &:after {
    content: '';
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    position: absolute;
    border: $plp-product-swatch-button-after-border;
    border-radius: $plp-product-swatch-radius;
    pointer-events: none;
    transition: $plp-product-swatch-button-selected-after-transition;
    transform: $plp-product-swatch-button-after-transform;
  }

  &.is-selected {
    &:after {
      border: $plp-product-swatch-button-selected-after-border;
      border-bottom: $plp-product-swatch-button-selected-after-border-bottom;
      top: $plp-product-swatch-button-selected-after-top;
      left: 0px;
      transform: $plp-product-swatch-button-selected-after-transform;
    }
  }

  &-link {
    box-sizing: border-box;
    display: inline-block;
    line-height: 0;
    overflow: hidden;
    width: 100%;
    height: 100%;
    border-radius: $plp-product-swatch-radius;
  }

  &--pdp {
    margin: 0;
  }

  &-linkImage {
    width: 44px;
    height: 44px;
    position: relative;
    top: -8px;
    left: -8px;
    user-select: none;
  }
}

@media $min-tablet {
  .Swatch {
    &--pdp {
      margin: 0;
      width: $SwatchSizeDesktop;
      height: $SwatchSizeDesktop;
    }
  }
}

@media $desktop {
  .Swatch {
    margin: $SwatchMarginDesktop;
    width: $SwatchSizeDesktop;
    height: $SwatchSizeDesktop;

    &:first-child {
      margin: $SwatchMarginDesktop-firstChild;
    }
  }
}
