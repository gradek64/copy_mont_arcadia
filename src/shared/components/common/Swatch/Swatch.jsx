import PropTypes from 'prop-types'
import React, { Component } from 'react'
import Image from '../Image/Image'
import { Link } from 'react-router'

export default class Swatch extends Component {
  static propTypes = {
    className: PropTypes.string,
    selected: PropTypes.bool,
    onSelect: PropTypes.func,
    index: PropTypes.number,
    imageUrl: PropTypes.string,
    colourName: PropTypes.string,
    seoUrl: PropTypes.string,
    onSwatchClick: PropTypes.func,
  }

  static defaultProps = {
    onSwatchClick: () => {},
  }

  onSelect = (event) => {
    event.preventDefault()
    const { onSelect, index, onSwatchClick } = this.props
    if (onSelect) onSelect(event, index)
    if (onSwatchClick) onSwatchClick(event)
  }

  render() {
    const {
      index,
      imageUrl,
      selected,
      colourName,
      className,
      seoUrl,
    } = this.props

    const colour = typeof colourName === 'string' && colourName.trim()
    const backgroundStyle = colour ? { background: `#${colour}` } : undefined
    // SeoUrl may be blank as the API is not ready. So we back it up with '' to make sure Link doesn't break
    // It is over written with a back up click handler, but will currently break in no-js mode, but this isn't a priority.
    return (
      <div
        className={`Swatch${className ? ` ${className}` : ''}${
          selected ? ' is-selected' : ''
        }`}
      >
        <Link
          key={index}
          className={`Swatch-link${selected ? ' is-selected' : ''}`}
          to={seoUrl || ''}
          onClick={this.onSelect}
          style={backgroundStyle}
        >
          {!colour && <Image className="Swatch-linkImage" src={imageUrl} />}
        </Link>
      </div>
    )
  }
}
