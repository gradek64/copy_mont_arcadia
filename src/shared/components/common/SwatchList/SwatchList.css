/** @define SwatchList */

.SwatchList {
  display: flex;
  justify-content: $plp-product-swatch-list-justify;
  align-items: center;
  padding: $plp-product-swatch-list-padding;
  margin-top: $plp-product-swatch-list-margin-top;
  margin-left: $plp-product-swatch-list-margin-left;

  &-button {
    background: $plp-product-swatch-button-image no-repeat;
    background-size: $plp-product-swatch-button-image-size;
    background-position: center center;
    display: inline-block;
    width: 22px;
    height: 22px;
    font-size: $plp-product-swatch-button-font-size;
    line-height: 1.1;
    vertical-align: middle;
    text-transform: $plp-product-swatch-button-text-transform;
    text-decoration: $plp-product-swatch-button-text-decoration;
    color: $plp-product-swatch-button-font-color;
    -webkit-tap-highlight-color: transparent;
    transform: translate3d(0, 0, 0);
    cursor: pointer;
    user-select: none;

    &--next {
      margin: $plp-product-swatch-button-next-margin;
    }
    &--prev {
      margin: $plp-product-swatch-button-prev-margin;
      transform: $plp-product-swatch-button-prev-transform;
    }
  }
  &-row {
    display: flex;
    flex-direction: row;
    width: 100%;
    justify-content: center;
    margin-bottom: 15px;
  }
  &-cell {
    width: 28px;
    margin: 0 10px;
    justify-content: center;
  }
  &--pdp {
    justify-content: flex-start;
    padding: $pdp-product-swatch-list-padding;

    & .SwatchList-row {
      flex-wrap: wrap;
      margin-bottom: $pdp-product-swatch-row-margin-bottom;
    }

    & .SwatchList-cell {
      margin-bottom: $pdp-product-swatch-margin-bottom;
    }
  }
}

@media $min-tablet {
  .SwatchList {
    &-button {
      background-size: $plp-product-swatch-button-image-size-tablet;
      width: 28px;
      height: 28px;
    }
  }
}
