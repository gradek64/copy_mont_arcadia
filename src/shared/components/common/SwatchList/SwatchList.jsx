import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router'
import { connect } from 'react-redux'
import Swatch from '../Swatch/Swatch'
import * as swatchesActions from '../../../actions/common/swatchesActions'
import { getRouteFromUrl } from '../../../lib/get-product-route'

@connect(
  (state) => ({
    location: state.routing.location,
  }),
  swatchesActions
)
class SwatchList extends Component {
  static propTypes = {
    productId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    swatches: PropTypes.array,
    selected: PropTypes.number,
    maxSwatches: PropTypes.number,
    page: PropTypes.number,
    nextSwatches: PropTypes.func,
    prevSwatches: PropTypes.func,
    pageClass: PropTypes.string,
    onSelect: PropTypes.func,
    seoUrl: PropTypes.string,
    resetSwatchesPage: PropTypes.func,
    onSwatchClick: PropTypes.func,
  }

  static contextTypes = {
    l: PropTypes.func,
  }

  getPrevButton = (productId) => {
    const { l } = this.context
    const { seoUrl } = this.props
    return (
      <Link
        key={`prev-${productId}`}
        className="SwatchList-button SwatchList-button--prev"
        onClick={this.prevSwatches.bind(this, productId)} // eslint-disable-line react/jsx-no-bind
        to={getRouteFromUrl(seoUrl)}
      >
        {l`See Prev`}
      </Link>
    )
  }

  getNextButton = (productId) => {
    const { l } = this.context
    const { seoUrl } = this.props
    return (
      <Link
        key={`next-${productId}`}
        className="SwatchList-button SwatchList-button--next"
        onClick={this.nextSwatches.bind(this, productId)} // eslint-disable-line react/jsx-no-bind
        to={getRouteFromUrl(seoUrl)}
      >
        {l`See More`}
      </Link>
    )
  }

  getSwatch(swatch, index) {
    const {
      onSelect,
      pageClass,
      selected: selectedIndex,
      onSwatchClick,
    } = this.props
    const props = {
      key: index,
      index,
      selected: selectedIndex === index,
      onSelect,
      onSwatchClick,
    }
    return (
      <Swatch
        className={pageClass ? `Swatch--${pageClass}` : ''}
        {...props}
        {...swatch}
      />
    )
  }

  getSwatchesInRows(swatches) {
    return (
      <div className="SwatchList-row">
        {swatches.map((swatch, index) => this.getSwatchInCell(swatch, index))}
      </div>
    )
  }

  getSwatchInCell(swatch, selected) {
    return (
      <div className="SwatchList-cell" key={selected}>
        {swatch && this.getSwatch(swatch, selected)}
      </div>
    )
  }

  prevSwatches(productId, e) {
    e.preventDefault()
    e.stopPropagation()
    this.props.prevSwatches(productId)
  }

  nextSwatches(productId, e) {
    e.preventDefault()
    e.stopPropagation()
    this.props.nextSwatches(productId)
  }

  getMaxOffset = (numSwatches, maxSwatches) =>
    numSwatches > maxSwatches ? numSwatches - maxSwatches : 0

  createSwatchList = ({
    swatches,
    offset,
    beginSliceIndex,
    endSliceIndex,
    addPrevBtn,
    addNextBtn,
    productId,
    maxOffset,
  }) => {
    // Create a new list with or without a previous button
    const list = addPrevBtn ? [this.getPrevButton(productId)] : []
    // Add a next button if required
    const nextBtn = addNextBtn && maxOffset ? this.getNextButton(productId) : []
    // If swatches length is more than "maxOffset" slice swatches
    const slice = maxOffset ? [beginSliceIndex, endSliceIndex] : []
    // Apply "swatchOffset" if add previous button is true
    const swatchOffset = addPrevBtn ? offset + 1 : offset

    return list
      .concat(
        swatches
          .slice(...slice)
          .map((swatch, i) => this.getSwatch(swatch, swatchOffset + i))
      )
      .concat(nextBtn)
  }

  renderSwatches(
    swatches,
    maxSwatches,
    productId,
    offset,
    pageClass,
    resetSwatchesPage
  ) {
    // If component is loaded on PDP, no carousel.
    const noPaging = swatches.length === maxSwatches
    if (pageClass === 'pdp' && noPaging) {
      return this.getSwatchesInRows(swatches)
    }

    // The width of the parent container can not fit the number of swatches a
    // carousel is implemented starting from here.
    const { length: numSwatches } = swatches
    const maxOffset = this.getMaxOffset(numSwatches, maxSwatches)
    const args = {
      swatches,
      offset,
      beginSliceIndex: 0,
      endSliceIndex: 0,
      addPrevBtn: false,
      addNextBtn: false,
      productId,
      maxOffset,
    }

    // PATCH -
    // Changing the grid layout and resizing the browser will break the carousel.
    // Reset the swatches page back to zero will fix this issue
    if (maxSwatches + offset > numSwatches && maxSwatches < numSwatches) {
      resetSwatchesPage()
    }

    // Start of the swatch list, add next button
    if (offset === 0) {
      args.endSliceIndex = maxSwatches - 1
      args.addNextBtn = true
    }

    // Add previous button
    if (offset > 0) {
      args.beginSliceIndex = offset + 1
      args.endSliceIndex = args.beginSliceIndex + maxSwatches - 2
      args.addPrevBtn = true
      args.addNextBtn = true

      // End of the swatch list, remove next button.
      if (offset === maxOffset) {
        args.beginSliceIndex = offset + 1
        args.endSliceIndex = offset + maxSwatches
        args.addNextBtn = false
      }
    }

    return this.createSwatchList(args)
  }

  render() {
    const {
      swatches,
      maxSwatches,
      productId,
      page: offset,
      pageClass,
      resetSwatchesPage,
    } = this.props
    return (
      <div
        className={`SwatchList${pageClass ? ` SwatchList--${pageClass}` : ''}`}
      >
        {swatches &&
          swatches.length > 0 &&
          this.renderSwatches(
            swatches,
            maxSwatches,
            productId,
            offset,
            pageClass,
            resetSwatchesPage
          )}
      </div>
    )
  }
}

export default SwatchList
