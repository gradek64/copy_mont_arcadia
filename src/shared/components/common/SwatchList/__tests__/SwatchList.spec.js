import testComponentHelper from 'test/unit/helpers/test-component'
import SwatchList from '../SwatchList'
import Swatch from '../../Swatch/Swatch'

const renderComponent = testComponentHelper(SwatchList.WrappedComponent)

const swatches = [
  {
    imageUrl: 'swatch0',
    colourName: 'swatch0',
    swatchProduct: {
      productId: 'swatch0',
    },
  },
  {
    imageUrl: 'swatch1',
    colourName: 'swatch1',
    swatchProduct: {
      productId: 'swatch1',
    },
  },
  {
    imageUrl: 'swatch2',
    colourName: 'swatch2',
    swatchProduct: {
      productId: 'swatch2',
    },
  },
]

const lotsOfSwatches = [
  ...swatches,
  {
    imageUrl: 'swatch3',
    colourName: 'swatch3',
    swatchProduct: {
      productId: 'swatch3',
    },
  },
  {
    imageUrl: 'swatch4',
    colourName: 'swatch4',
    swatchProduct: {
      productId: 'swatch4',
    },
  },
  {
    imageUrl: 'swatch5',
    colourName: 'swatch5',
    swatchProduct: {
      productId: 'swatch5',
    },
  },
  {
    imageUrl: 'swatch6',
    colourName: 'swatch6',
    swatchProduct: {
      productId: 'swatch6',
    },
  },
  {
    imageUrl: 'swatch7',
    colourName: 'swatch7',
    swatchProduct: {
      productId: 'swatch7',
    },
  },
]

const evenMoreSwatches = [
  ...lotsOfSwatches,
  {
    imageUrl: 'swatch8',
    colourName: 'swatch8',
    swatchProduct: {
      productId: 'swatch8',
    },
  },
  {
    imageUrl: 'swatch9',
    colourName: 'swatch9',
    swatchProduct: {
      productId: 'swatch9',
    },
  },
  {
    imageUrl: 'swatch10',
    colourName: 'swatch10',
    swatchProduct: {
      productId: 'swatch10',
    },
  },
  {
    imageUrl: 'swatch11',
    colourName: 'swatch11',
    swatchProduct: {
      productId: 'swatch11',
    },
  },
  {
    imageUrl: 'swatch12',
    colourName: 'swatch12',
    swatchProduct: {
      productId: 'swatch12',
    },
  },
]

const props = {
  seoUrl: '',
  swatches,
  page: 0,
  selected: 0,
  maxSwatches: 4,
  productId: '1234',
  resetSwatchesPage: jest.fn(),
}

describe('<SwatchList/>', () => {
  describe('@renders', () => {
    it('renders in default state', () => {
      expect(renderComponent().getTree()).toMatchSnapshot()
    })

    it('should render 3 <Swatch/> components if 3 swatches have been passed in via `swatches` props', () => {
      const { wrapper } = renderComponent({ ...props })
      expect(wrapper.find(Swatch).length).toEqual(3)
      expect(renderComponent(props).getTree()).toMatchSnapshot()
    })

    it('correct swatch is selected', () => {
      const { wrapper } = renderComponent({ ...props, selected: 1 })
      expect(
        wrapper
          .find(Swatch)
          .at(1)
          .prop('selected')
      ).toBe(true)
    })

    it('displays a next page button when needed', () => {
      const { wrapper } = renderComponent({
        ...props,
        swatches: lotsOfSwatches,
        selected: 1,
      })
      expect(wrapper.find('.SwatchList-button').length).toEqual(1)
    })

    it('has no buttons if not needed', () => {
      const { wrapper } = renderComponent({ ...props, selected: 1 })
      expect(wrapper.find('.SwatchList-button').length).toEqual(0)
    })

    it('always shows correct amount of swatches on first page when there are pages', () => {
      const { wrapper } = renderComponent({
        ...props,
        swatches: evenMoreSwatches,
        selected: 1,
        page: 0,
        maxSwatches: 6,
      })

      const length2 = wrapper.find(Swatch).length

      expect(length2).toEqual(5)
    })

    it('always shows correct amount of swatches on second page when there are pages', () => {
      const { wrapper } = renderComponent({
        ...props,
        swatches: evenMoreSwatches,
        selected: 1,
        page: 1,
        maxSwatches: 6,
      })

      const length2 = wrapper.find(Swatch).length

      expect(length2).toEqual(4)
    })

    it('shows swatches in rows on a PDP, with a final short row when there is no pagination', () => {
      const { wrapper } = renderComponent({
        ...props,
        pageClass: 'pdp',
        swatches: evenMoreSwatches,
        selected: 1,
        maxSwatches: 13,
      })

      const numSwatches = wrapper.find(Swatch).length
      expect(wrapper.find('.SwatchList-row').length).toEqual(1)
      expect(wrapper.find('.SwatchList-cell').length).toEqual(13)
      expect(numSwatches).toEqual(13)
    })

    it('should not show swatches in rows on a PLP', () => {
      const { wrapper } = renderComponent({
        ...props,
        pageClass: 'plp',
        swatches: evenMoreSwatches,
        selected: 1,
        maxSwatches: 13,
      })

      const numSwatches = wrapper.find(Swatch).length
      expect(wrapper.find('.SwatchList-row').length).toBe(0)
      expect(wrapper.find('.SwatchList-row--short').length).toBe(0)
      expect(wrapper.find('.SwatchList-cell').length).toBe(0)
      expect(wrapper.find('.SwatchList-cell-short').length).toBe(0)
      expect(numSwatches).toEqual(13)
    })

    it('shows swatches in rows on a PDP when there is no pagination', () => {
      const evenNumberOfSwatches = evenMoreSwatches.splice(0, 12)
      const { wrapper } = renderComponent({
        ...props,
        pageClass: 'pdp',
        swatches: evenNumberOfSwatches,
        selected: 1,
        maxSwatches: 12,
      })

      const numSwatches = wrapper.find(Swatch).length
      expect(wrapper.find('.SwatchList-row').length).toEqual(1)
      expect(wrapper.find('.SwatchList-cell').length).toEqual(12)
      expect(wrapper.find('.SwatchList-cell--short').length).toEqual(0)
      expect(numSwatches).toEqual(12)
    })

    it('displays a next page button when needed', () => {
      const { wrapper } = renderComponent({
        ...props,
        selected: 1,
        swatches: lotsOfSwatches,
        page: 0,
        maxSwatches: 4,
      })
      expect(wrapper.find('.SwatchList-button').length).toEqual(1)
      expect(wrapper.find('.SwatchList-button--next').length).toEqual(1)
    })

    it('displays a prev page button when needed', () => {
      const { wrapper } = renderComponent({
        ...props,
        selected: 1,
        swatches: lotsOfSwatches,
        page: 4,
        maxSwatches: 4,
      })
      expect(wrapper.find('.SwatchList-button').length).toEqual(1)
      expect(wrapper.find('.SwatchList-button--prev').length).toEqual(1)
    })

    it('displays both buttons when needed', () => {
      const { wrapper } = renderComponent({
        ...props,
        selected: 1,
        swatches: lotsOfSwatches,
        page: 1,
        maxSwatches: 6,
      })
      expect(wrapper.find('.SwatchList-button--prev').length).toEqual(1)
      expect(wrapper.find('.SwatchList-button--next').length).toEqual(1)
    })

    it('passes down `onSwatchClick` to Swatch', () => {
      const onSwatchClick = jest.fn()
      const { wrapper } = renderComponent({
        ...props,
        selected: 1,
        swatches: lotsOfSwatches,
        page: 1,
        onSwatchClick,
      })
      expect(
        wrapper
          .find(Swatch)
          .at(1)
          .props().onSwatchClick
      ).toEqual(onSwatchClick)
    })

    it('passes down `onSelect` click handler to Swatch', () => {
      const onSelect = jest.fn()
      const { wrapper } = renderComponent({
        ...props,
        selected: 1,
        swatches: lotsOfSwatches,
        page: 1,
        onSelect,
      })
      expect(
        wrapper
          .find(Swatch)
          .at(1)
          .props().onSelect
      ).toEqual(onSelect)
    })

    it('fires nextSwatches when next button is clicked', () => {
      const nextSwatches = jest.fn()
      const { wrapper } = renderComponent({
        ...props,
        selected: 1,
        swatches: lotsOfSwatches,
        page: 1,
        nextSwatches,
      })
      wrapper.find('.SwatchList-button--next').simulate('click', {
        preventDefault: () => {},
        stopPropagation: () => {},
      })
      expect(nextSwatches).toHaveBeenCalled()
    })

    it('fires `prevSwatches` when prev button is clicked', () => {
      const prevSwatches = jest.fn()
      const { wrapper } = renderComponent({
        ...props,
        selected: 1,
        swatches: lotsOfSwatches,
        page: 1,
        prevSwatches,
      })
      wrapper.find('.SwatchList-button--prev').simulate('click', {
        preventDefault: () => {},
        stopPropagation: () => {},
      })
      expect(prevSwatches).toHaveBeenCalled()
    })

    it('should reset swatches page when the page plus maxSwatches is greater than swatches.length', () => {
      const { instance } = renderComponent({
        ...props,
        selected: 1,
        swatches: lotsOfSwatches,
        maxSwatches: 4,
        page: 6,
      })

      expect(instance.props.resetSwatchesPage).toBeCalled()
    })
  })
})
