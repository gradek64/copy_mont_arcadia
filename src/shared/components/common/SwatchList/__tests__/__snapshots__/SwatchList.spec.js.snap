// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SwatchList/> @renders renders in default state 1`] = `
<div
  className="SwatchList"
/>
`;

exports[`<SwatchList/> @renders should render 3 <Swatch/> components if 3 swatches have been passed in via \`swatches\` props 1`] = `
<div
  className="SwatchList"
>
  <Swatch
    className=""
    colourName="swatch0"
    imageUrl="swatch0"
    index={0}
    key="0"
    onSwatchClick={[Function]}
    selected={true}
    swatchProduct={
      Object {
        "productId": "swatch0",
      }
    }
  />
  <Swatch
    className=""
    colourName="swatch1"
    imageUrl="swatch1"
    index={1}
    key="1"
    onSwatchClick={[Function]}
    selected={false}
    swatchProduct={
      Object {
        "productId": "swatch1",
      }
    }
  />
  <Swatch
    className=""
    colourName="swatch2"
    imageUrl="swatch2"
    index={2}
    key="2"
    onSwatchClick={[Function]}
    selected={false}
    swatchProduct={
      Object {
        "productId": "swatch2",
      }
    }
  />
</div>
`;
